### PR TITLE
xstyled v2

### PR DESCRIPTION
**Pull request by gregberge**
## Features

- [x] TypeScript typings
- [x] Emotion v11
- [x] Get rid of `forwardedAs` (use `as` instead)
- [x] Allow configuration of `rootFontSize` for `rpx` utilities
- [x] Introduce a `defaultTheme`
- [x] Add state support (`motionSafe`, `motionReduce`, `first`, `last`, `odd`, `even`, `visited`, `checked`, `focusWithin`, `hover`, `focus`, `focusVisible`, `active`, `disabled`, `placeholder`)
- [x] Support shorthand styles
- [x] Add `gridTemplateColumns` and `gridTemplateRows` support in theme
- [x] Introduce default value if value is `true`
- [x] Support string value as number for pixels and durations
- [x] Support `<Box.{element}>` syntax
- [x] Add `transitionProperties` and `timingFunctions` support in theme
- [x] Add `inset` support in theme
- [x] Support arrays for list properties in theme
- [x] Add `getDuration` getter
- [x] Add `Preflight`: a set of global style to start a project
- [x] Support default value in `th` and `th.*` 
- [x] Allow to create complex style generator using `style` utility
- [x] Add `aliasColor`and `generateHexAlphaVariants` utilities
- [x] Add `useTh` and other hooks to get theme props
- [x] Add `x.extend` and `createX` utilities 

## Utilities

- [x] Add svg utilities: `fill`, `stroke`
- [x] Add space utilties: `spaceX`, `spaceX`
- [x] Add sizing utilities: `w`, `h`
- [x] Add lists utilities: `listStyleType`, `listStyleTypePosition`
- [x] Add borders utilities: `outline`, `divideX`, `divideY`, `divideXReverse`, `divideYReverse`, `divideStyle`, `divideColor`, `ring`, `ringInset`, `ringColor`
- [x] Add typography utilities: `textDecoration`, `textOverflow`, `whiteSpace`
- [x] Add layout utilities: `overflowX`, `overflowY`, `boxSizing`, `container`, `visibility`, `overscrollBehavior`, `objectFit`
- [x] Add grid utilities: `gap`
- [x] Add tables utilities: `tableLayout`, `borderCollapse`
- [x] Add transitions utilities: `transitionProperty`, `transitionDuration`, `transitionTimingFunction`, `transitionDelay`
- [x] Add animations utilities: `animation`
- [x] Add interactivity utilities: `appearance`, `cursor`, `pointerEvents`, `resize`, `userSelect`
- [x] Add transforms utilities: `transform`, `transformOrigin`, `translateX`, `translateY`,  `rotate`, `skewX`, `skewY`, `scale`, ` scaleX`, `scaleY`
- [x] Add backgrounds utilities: `backgroundAttachment`, `backgroundClip`, `gradientFrom`, `gradientVia`, `gradientTo` 

## Breaking changes

- Emotion v10 is no longer supported
- Remove default `space`
- Remove default `breakpoints`
- Rename `breakpoints` theme section to `screens`
- Remove `variant` utility
- Remove `width` and `height` utility (only aliases remain `w` and `h`)
- Replace `gridGap` by `gap`
- Remove `border{direction}` and `border{direction}Color` utilities
- Reorganize all utilities

### Mixed

- [x] Homepage
- [x] Check Customzing section of all pages
- [x] Sort navigation
- [x] Mention https://github.com/smooth-code/xstyled/issues/94#issuecomment-748413963 in docs
- [x] Add colors everywhere
- [x] Check all features in Emotion & Styled Components
- [x] Fix tests

### Smooth DOC

- [x] Integrate Preview Live Editor
- [x] Make preview full size with no margin
- [v2.1] DocSearch v2
- [x] Reduce nav font size
- [x] Fix nav scrolling issue

### Getting Started

- [x] Introduction
- [x] Installation
- [x] Release Notes -> GitHub Release Page
- [x] Upgrade Guide
- [x] Migrate from Styled Components
- [x] Migrate from Emotion
- [x] Browser Support

### Core Concepts

- [x] Utility-First
- [x] Responsive Design
- [x] Hover, Focus & Other States
- [x] Dark Mode
- [x] Adding Base Styles
- [v2.1] Extracting Components
- [x] Adding Utilities
- [x] Composing Utilities
- [x] Styled Components 
- [x] Hooks
- [x] TypeScript
- [x] PropTypes
- [x] Transformers

### Customization

- [x] Theme
- [x] Colors

### Layout

- [x] Container
- [x] Box Sizing
- [x] Display
- [v2.1] Floats
- [v2.1] Clear
- [x] Object Fit
- [v2.1] Object Position
- [x] Overflow
- [x] Overscroll Behavior
- [x] Position
- [x] Top / Right / Bottom / Left
- [x] Visibility
- [x] Z-Index

### Flexbox

- [x] Flex Direction
- [x] Flex Wrap
- [x] Flex
- [x] Flex Grow
- [x] Flex Shrink
- [x] Order
- [x] Grid System

### Grid

- [x] Grid Template Columns
- [x] Grid Column Start / End
- [x] Grid Template Rows
- [x] Grid Row Start / End
- [x] Grid Template Areas
- [x] Grid Auto Flow
- [x] Grid Auto Columns
- [x] Grid Auto Rows
- [x] Gap

### Box Alignment

- [x] Justify Content
- [x] Justify Items
- [x] Justify Self
- [x] Align Content
- [x] Align Items
- [x] Align Self
- [v2.1] Place Content
- [v2.1] Place Items
- [v2.1] Place Self

### Spacing

- [x] Padding
- [x] Margin
- [x] Space Between

### Sizing

- [x] Width
- [x] Min-Width
- [x] Max-Width
- [x] Height
- [x] Min-Height
- [x] Max-Height

### Typography

- [x] Font Family
- [x] Font Size
- [v2.1] Font Smoothing
- [x] Font Style
- [x] Font Weight
- [v2.1] Font Variant Numeric
- [x] Letter Spacing
- [x] Line Height
- [x] List Style Type
- [x] List Style Position
- [x] Text Align
- [x] Text Color
- [x] Text Decoration
- [x] Text Transform
- [x] Text Overflow
- [x] Vertical Align
- [x] Whitespace
- [v2.1] Word Break

### Backgrounds

- [x] Background Attachment
- [x] Background Clip
- [x] Background Color
- [x] Background Position
- [x] Background Repeat
- [x] Background Size
- [x] Background Image
- [x] Gradient Color Stops

### Borders

- [x] Border Radius
- [x] Border Width
- [x] Border Color
- [x] Border Style
- [x] Divide Width
- [x] Divide Color
- [x] Divide Style
- [x] Ring Width
- [x] Ring Color

### Effects

- [x] Box Shadow
- [x] Opacity

### Tables

- [x] Border Collapse
- [x] Table Layout

### Transitions

- [x] Transition
- [x] Transition Property
- [x] Transition Duration
- [x] Transition Timing Function
- [x] Transition Delay

### Animations

- [x] Animation

### Transforms

- [x] Transform
- [x] Transform Origin
- [x] Scale
- [x] Rotate
- [x] Translate
- [x] Skew

### Interactivity

- [x] Appearance
- [x] Cursor
- [x] Outline
- [x] Pointer Events
- [x] Resize
- [x] User Select

### Accessibility

- [v2.1] Screen Readers

### CODE REVIEW

**Comment by yairEO** [comment]
image

`gap` is not just for `grid` but also for flexbox

https://web.dev/flexbox-gap

I would like to ask when will theme (`space` key) support will arrive for the CSS `gap` property?

File: website/pages/docs/getting-started/upgrade-guide.mdx [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+<carbon-ad />
+
+xstyled v2 is the new major version since v1 of xstyled released in May 2019.
+
+We know that xstyled is in the core of your project and that the migration could be difficult. For that purpose, we minimized the number of changes between the two versions, there is only few breaking changes.
+
+## New philosophy
+
+The v2 of xstyled favoured the usage of the `x` instead of the `styled` approach. Even if the `x` approach is always supported, it is better to use `x` as much as possible!
```
Comment (mariusbrn) [inline website/pages/docs/getting-started/upgrade-guide.mdx:20]: Even if the `Box` approach ?

**Comment by yairEO** [comment]
you stated here: 
> Rename breakpoints theme section to screens

Yet, it is not so in the dods:

https://xstyled.dev/docs/responsive-design

image

Also, there is no example to be found in the whole docs on how to use this in the CSS

**Comment by iManu** [comment]
Hello @gregberge, did you forget `ObjectPosition` here (notified for 2.1) or am I wrong ?

**Pull request merged**